### ISSUE TITLE
WebView2 initialization in headless mode not work when in invoked by windows service

### ISSUE THREAD

**Issue by fabioDG89** [MicrosoftEdge/WebView2Feedback#3458]
Hi,
i'm trying to use WebView2 in headless mode to export in pdf a page but when is invoked by a Windows Service doesn't work.

This is the scenario.

I have a web application named "WebAppService" in ASP.NET Core 6.0. running an html page.

```c#
var options = new WebApplicationOptions() {
    Args = args,
    ContentRootPath = WindowsServiceHelpers.IsWindowsService() ? AppContext.BaseDirectory : default
};

var builder = WebApplication.CreateBuilder(options);

builder.Services.AddControllers();

if (!Debugger.IsAttached) {
    builder.Host.UseWindowsService();
}

var app = builder.Build();
app.UseStaticFiles();
app.UseRouting();
app.UseEndpoints(endpoints => { endpoints.MapControllers(); });


app.Run("http://localhost:5000");
```

In the HTML page there is a button and on click i call a web API that call the executable used for the initialization of WebView2 and the export in pdf.

```c#  
[Route("[controller]")]
    public class ExportWebViewController:Controller {

        [HttpGet("ExecProcess")]
        public ActionResult ExecProcess() {
            Process.Start("C:\\Sources\\WebView2Headless\\ConsoleApp1\\bin\\Debug\\net6.0-windows\\ConsoleApp1.exe");
            return Ok();
        }
    }
```

The executable invoked "ConsoleApp1" contains the methods to initialize WebView2 in headless mode, navigate and export in pdf.

```c#
static void CreateWebView2() {
    System.Diagnostics.Debugger.Launch();
    CoreWebView2Controller browserController;
    string WEBVIEW2_RUNTIME_FOLDER = @"C:\Sources\WebView2Headless\ConsoleApp1\bin\Debug\WebView2Runtime.110.0.1587.63.x64";
    IntPtr HWND_MESSAGE = new IntPtr(-3);
    var uiThread = new Thread(() => {

        Dispatcher.CurrentDispatcher.Invoke(async () => {
            var environment = await CoreWebView2Environment.CreateAsync(WEBVIEW2_RUNTIME_FOLDER, null, null);

            browserController = await environment.CreateCoreWebView2ControllerAsync(HWND_MESSAGE);
            browserController.CoreWebView2.NavigationCompleted += NavigationComplete;

            browserController.CoreWebView2.Navigate("https://developer.microsoft.com/it-it/microsoft-edge/webview2/");

        });

        Dispatcher.Run();
    });

    uiThread.SetApartmentState(ApartmentState.STA);
    uiThread.Start();
    uiThread.Join();

    void NavigationComplete(object sender, CoreWebView2NavigationCompletedEventArgs args) {
        browserController.CoreWebView2.PrintToPdfAsync(@"C:\Temp\ExportFromWebView2.pdf");
        Console.WriteLine("Export complete!");
    }
}


[STAThread]
static void Main(string[] args) {
    Console.WriteLine("START -> CreateWebView2 Environment...");
    CreateWebView2();
}

Main(args);
```

When the web application "WebAppService" is running in debug mode in Visual Studio (or invoked by a Console Application), the export works.

When the web application is invoked by a Windows Service, the export doesn't work: the "ConsoleApplication1" is executed, but it stucks at method "CreateCoreWebView2ControllerAsync" (no exceptions are returned).

I'm using WebView2Runtime.110.0.1587.63.x64

Thanks for your help!

Fabio

**Comment by LiangTheDev** [comment]
Could this be related to https://github.com/MicrosoftEdge/WebView2Feedback/issues/1907? That's an unsupported scenario.

**Comment by fabioDG89** [comment]
No, in my scenario the windows service run under a user that is administrator

**Comment by LiangTheDev** [comment]
Then that could be due to difficulty to create non elevated WebView2 processes. WebView2 processes cannot run elevated. For normal session, we create non elevated WebView2 processes through explorer. I would imagine that it doesn't work when running in windows service. So, this is a scenario that we do not support currently.

**Comment by fabioDG89** [comment]
To avoid misunderstandings I explain my scenario:

This is the windows service that start the web application in ASP.NET Core 6.0.

![01](https://github.com/MicrosoftEdge/WebView2Feedback/assets/78020123/3e3499e5-7f68-4fe7-acee-cd4529f89b9e)

It is executed by the user "ServiceExecutor", that is local administrator:

![02](https://github.com/MicrosoftEdge/WebView2Feedback/assets/78020123/d2dc46e8-b2ff-4676-a897-5f52210b5a98)

Clicking on the button Export, it calls the action "ExecProcess" that execute the console application "ConsoleApp1.exe" that initialize WebView2 in headless mode, navigate and export in pdf.

![03](https://github.com/MicrosoftEdge/WebView2Feedback/assets/78020123/2e7aefb8-7c9c-4284-a8d8-47f1cfaeb235)


The process started by the user "ServiceExecutor", but nothing it is exported.

![04](https://github.com/MicrosoftEdge/WebView2Feedback/assets/78020123/ebf5a6af-8368-4087-ad4f-49e79ea5b53d)

This scenario could be supported?

Thanks for support!

**Comment by LiangTheDev** [comment]
This is a setup WebView2 has not been tested with, and therefore not officially supported.

In the setup, ConsoleApp1.exe is running elevated and as different user, probably in an invisible session. We have several hurdles to overcome: launch de-elevated WebView2 process, which depends on explorer, allow the launched WebView2 process running as different user (the same interactive user as explorer) to access connection end point created by the elevated process, and Chromium's sandbox code works in this invisible session.

That being said, WebView2 code might have already overcome the first 2 hurdles and it is the Chromium's sandbox code that is preventing it from working. You could try to create WebView with --no-sandbox additional browser arguments (`CoreWebView2Environment.CreateAsync(WEBVIEW2_RUNTIME_FOLDER, null, new CoreWebView2EnvironmentOptions("--no-sandbox");`) and see whether it works.

If it doesn't work, it is the first 2 hurdles that is breaking the scenario. Try to use latest version of Edge WebView2 runtime and see whether it works. We made some changes related to elevated app in Edge 112, which might or might not help. If still does not work, could you try to config it to run as a non admin user?

**Comment by LiangTheDev** [comment]
If it works with non admin but not with admin even using --no-sandbox, it means that dependency on explorer is the one that prevents it from working. Therefore, it is a WebView2 limitation with no workaround. Running the service with non admin account is the only workaround.
